src/github.com/matrix-org/dendrite/roomserver/input/events.go

```diff
@@ -103,13 +107,84 @@ func processRoomEvent(db RoomEventDatabase, ow OutputRoomEventWriter, input api.
-	//  * Write a message to the output logs containing:
-	//      - The event itself
-	//      - The visiblity of the event, i.e. who is allowed to see the event.
-	//      - The changes to the current state of the room.
 	return nil
 }
+
+func processInviteEvent(db RoomEventDatabase, ow OutputRoomEventWriter, input api.InputInviteEvent) (err error) {
+	if input.Event.StateKey() == nil {
+		return fmt.Errorf("invite must be a state event")
+	}
+
+	roomID := input.Event.RoomID()
+	targetUserID := *input.Event.StateKey()
+
+	updater, err := db.MembershipUpdater(roomID, targetUserID)
+	if err != nil {
+		return err
+	}
+
+	return withTransaction(updater, func() error {
+
+		if updater.IsJoin() {
+			// If the user is joined to the room then that takes precedence over this
+			// invite event. It makes little sense to move a user that is already
+			// joined to the room into the invite state.
+			// This could plausibly happen if an invite request raced with a join
+			// request for a user. For example if a user was invited to a public
+			// room and they joined the room at the same time as the invite was sent.
+			// The other way this could plausibly happen is if an invite raced with
+			// a kick. For example if a user was kicked from a room in error and in
+			// response someone else in the room re-invited them then it is possible
+			// for the invite request to race with the leave event so that the
+			// target receives invite before it learns that it has been kicked.
+			// There are a few ways this could be plausibly handled in the roomserver.
+			// 1) Store the invite, but mark it as retired. That will result in the
+			//    permanent rejection of that invite event. So even if the target
+			//    user leaves the room and the invite is retransmitted it will be
+			//    ignored. However a new invite with a new event ID would still be
+			//    accepted.
+			// 2) Silently discard the invite event. This means that if the event
+			//    was retransmitted at a later date after the target user had left
+			//    the room we would accept the invite. However since we hadn't told
+			//    the sending server that the invite had been discarded it would
+			//    have no reason to attempt to retry.
+			// 3) Signal the sending server that the user is already joined to the
+			//    room.
+			// For now we will implement option 2. Since in the abesence of a retry
+			// mechanism it will be equivalent to option 1, and we don't have a
+			// signalling mechanism to implement option 3.
+			return nil
+		}
+
+		outputUpdates, err := updateToInviteMembership(updater, &input.Event, nil)
+		if err != nil {
+			return err
+		}
+
+		return ow.WriteOutputEvents(roomID, outputUpdates)
+	})
+}
+
+// withTransaction runs a function inside a transaction.
+// Commits the transaction if the function succeeds.
+// Rollsback the transaction if the function fails or panics.
+func withTransaction(t types.Transaction, f func() error) (err error) {
```

Comment:
well, I'm certainly not in favour of having two copies of the code, nor am I in favour of having two functions which do similar but different things whose names differ only by case. There's nothing specific to the roomserver here; it belongs in `common` just as much as the other `WithTransaction` does.

The main difference seems to be that `common.WithTransaction` starts the transaction, whereas this one assumes you already have a Transaction. I think this might be more appropriately named `common.CompleteTransaction` or something; and you could implement `common.WithTransaction` in terms of it?